src/Compilers/Core/Desktop/CommandLine/CommonCompiler.CompilerEmitStreamProvider.cs

```diff
@@ -0,0 +1,121 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal abstract partial class CommonCompiler
+    {
+        /// <summary>
+        /// This implementation of <see cref="Compilation.EmitStreamProvider"/> will delay the creation
+        /// of the PE + PDB file until the compiler determines the compilation has succeeded.  This prevents
+        /// the compiler from deleting output from the previous compilation when a new compilation 
+        /// fails.
+        /// </summary>
+        private sealed class CompilerEmitStreamProvider : Compilation.EmitStreamProvider, IDisposable
```

Comment:
@gafter nailed the reason I made this a nested type.  No one else needs to use it hence make it nested to reduce namespace pollution.